File: iup/html/en/7gui/circle_drawer.c
#include <stdlib.h>
#include <math.h>
#include <iup.h>
#include "iupdraw.h"

enum { CREATE, REMOVE, CHANGE_RADIUS };

typedef struct _Circle{
  int x;
  int y;
  int r;
  int rOriginal;
  int rLast;
  int id;
} Circle;

typedef struct _Operation{
  int type;
  Circle circle;
} Operation;

#define DEF_RADIUS 30
#define MAX 80

static Circle circles[MAX];
static int circlesCount = 0;
static int lastId = 0;
static int highlightedCircle = -1;

static Operation undoRedoList[MAX];
static int operationCount = 0;
static int currentOperationId = 0;

static void addOperation(int type, Circle circle)
{
  if (currentOperationId < operationCount - 1)
    operationCount = currentOperationId + 1;
  undoRedoList[operationCount].type = type;
  undoRedoList[operationCount].circle = circle;
  currentOperationId = operationCount;
  ++operationCount;
}

static void undo()
{
  Operation currentOperation = undoRedoList[currentOperationId];

  if (currentOperationId < 0)
    return;

  switch (currentOperation.type)
  {
    case CREATE:
      remove(currentOperation.circle.id);
      break;
    case REMOVE:
      insert(currentOperation.circle);
      break;
    case CHANGE_RADIUS:
      update(currentOperation.circle.id, currentOperation.circle.rLast);
      break;
  }
  currentOperationId--;
}

static void redo()
{
  Operation currentOperation;

  if (currentOperationId + 1 == operationCount)
    return;

  currentOperation = undoRedoList[currentOperationId + 1];

  switch (currentOperation.type)
  {
    case CREATE:
      insert(currentOperation.circle);
      break;
    case REMOVE:
      remove(currentOperation.circle.id);
      break;
    case CHANGE_RADIUS:
      update(currentOperation.circle.id, currentOperation.circle.r);
      break;
  }
  currentOperationId++;
}

static int undoLeft()
{
  if (currentOperationId < 0)
    return 0;

  return 1;
}

static int redoLeft()
{
  if (currentOperationId + 1 == operationCount)
    return 0;

  return 1;
}

static int getCount()
{
  return circlesCount;
}

static int getLasId()
{
  return lastId;
}

static Circle get(int id)
{
  int i;
  for (i = 0; i < circlesCount; i++)
  {
    if (circles[i].id == id)
      break;
  }
  return circles[i];
}

static int create(int x, int y, int r)
{
  Circle circle;
  if (circlesCount >= 50)
    return 0;
  circles[circlesCount].x = x;
  circles[circlesCount].y = y;
  circles[circlesCount].r = r;
  circles[circlesCount].rOriginal = r;
  circles[circlesCount].rLast = r;
  circles[circlesCount].id = lastId;
  ++circlesCount;
  ++lastId;
  return lastId - 1;
}

static int insert(Circle circle)
{
  int i;
  if (circlesCount >= 50)
    return 0;
  for (i = circlesCount - 1; i >= 0; i--)
  {
    if (circle.id > circles[i].id)
      break;
    circles[i + 1] = circles[i];
  }
  circles[i + 1] = circle;
  ++circlesCount;
  return 1;
}

static int remove(int id)
{
  int i, j;
  for (i = 0; i < circlesCount; i++)
  {
    if (circles[i].id == id)
      break;
  }

  for (j = i; j < circlesCount; j++)
    circles[j] = circles[j + 1];
  --circlesCount;
  return 1;
}

static int update(int id, int r)
{
  int i;
  for (i = 0; i < circlesCount; i++)
  {
    if (circles[i].id == id)
      break;
  }
  circles[i].rLast = circles[i].r;
  circles[i].r = r;
  return 1;
}

static int pick(int x, int y)
{
  int i;
  double dist, dx, dy;
  int selectedCircle = -1;
  for (i = 0; i < circlesCount; i++)
  {
    double dx = circles[i].x - x;
    double dy = circles[i].y - y;
    double d = sqrt((dx*dx) + (dy*dy));
    if (d < circles[i].r)
    {
      if (selectedCircle < 0)
      {
        selectedCircle = circles[i].id;
        dist = d;
      }
      else if (d < dist)
        selectedCircle = circles[i].id;
    }
  }

  return selectedCircle;
}

int bt_undo_cb(Ihandle *self)
{
  Ihandle *canvas = IupGetDialogChild(self, "CANVAS");
  Ihandle *undoButton = IupGetDialogChild(self, "UNDO");
  Ihandle *redoButton = IupGetDialogChild(self, "REDO");

  undo();

  if (undoLeft())
    IupSetAttribute(undoButton, "ACTIVE", "YES");
  else
    IupSetAttribute(undoButton, "ACTIVE", "NO");


  if (redoLeft())
    IupSetAttribute(redoButton, "ACTIVE", "YES");
  else
    IupSetAttribute(redoButton, "ACTIVE", "NO");

  IupUpdate(canvas);

  return IUP_DEFAULT;
}

int bt_redo_cb(Ihandle *self)
{
  Ihandle *canvas = IupGetDialogChild(self, "CANVAS");
  Ihandle *undoButton = IupGetDialogChild(self, "UNDO");
  Ihandle *redoButton = IupGetDialogChild(self, "REDO");

  redo();

  if (undoLeft())
    IupSetAttribute(undoButton, "ACTIVE", "YES");
  else
    IupSetAttribute(undoButton, "ACTIVE", "NO");


  if (redoLeft())
    IupSetAttribute(redoButton, "ACTIVE", "YES");
  else
    IupSetAttribute(redoButton, "ACTIVE", "NO");

  IupUpdate(canvas);

  return IUP_DEFAULT;
}
static int motion_cb(Ihandle *ih, int x, int y, char* status)
{
  highlightedCircle = pick(x, y);
  IupUpdate(ih);
  return IUP_DEFAULT;
}

static int button_cb(Ihandle *ih, int but, int pressed, int x, int y, char* status)
{
  Ihandle *dial = IupGetDialog(ih);

  if (but == IUP_BUTTON1 && pressed)
  {
    Ihandle *undoButton = IupGetDialogChild(ih, "UNDO");
    Ihandle *redoButton = IupGetDialogChild(ih, "REDO");
    int id = create(x, y, DEF_RADIUS);
    addOperation(CREATE, get(id));
    /***********************************************************************************/
    for (int i = 0; i < operationCount; i++)
    {
      printf("list[%d] = %d\n", i, undoRedoList[i].circle.id);
    }
    printf("Count = %d\n", operationCount);
    printf("Operation Ind = %d\n", operationCount);
    /***********************************************************************************/

    if (undoLeft())
      IupSetAttribute(undoButton, "ACTIVE", "YES");
    else
      IupSetAttribute(undoButton, "ACTIVE", "NO");

    if (redoLeft())
      IupSetAttribute(redoButton, "ACTIVE", "YES");
    else
      IupSetAttribute(redoButton, "ACTIVE", "NO");
  }
  else if (but == IUP_BUTTON3 && pressed)
  {
    int id = pick(x, y);
    if (id >= 0)
    {
      Ihandle *configDial = (Ihandle *)IupGetAttribute(dial, "CONFIGDIAL");
      IupSetInt(configDial, "CIRCLEID", id);
      IupPopup(IupGetAttribute(dial, "CONFIGDIAL"), IUP_CENTER, IUP_CENTER);
      highlightedCircle = -1;
    }
  }
  IupUpdate(ih);

  return IUP_DEFAULT;
}

static int action(Ihandle *ih, float posx, float posy)
{
  int i, w, h;

  IupDrawBegin(ih);

  IupDrawGetSize(ih, &w, &h);

  IupSetAttribute(ih, "DRAWCOLOR", "255 255 255");
  IupSetAttribute(ih, "DRAWSTYLE", "FILL");
  IupDrawRectangle(ih, 0, 0, w - 1, h - 1);

  IupSetAttribute(ih, "DRAWCOLOR", "0 0 0");
  IupSetAttribute(ih, "DRAWSTYLE", "STROKE");

  for (i = 0; i < getCount(); i++)
  {
    Circle circle = circles[i];

    if (highlightedCircle == circle.id)
    {
      IupSetAttribute(ih, "DRAWCOLOR", "128 128 128");
      IupSetAttribute(ih, "DRAWSTYLE", "FILL");
    }
    else
    {
      IupSetAttribute(ih, "DRAWCOLOR", "0 0 0");
      IupSetAttribute(ih, "DRAWSTYLE", "STROKE");
    }

    IupDrawArc(ih, circles[i].x - circles[i].r, circles[i].y - circles[i].r,
               circles[i].x + circles[i].r, circles[i].y + circles[i].r, 0., 360.);
  }

  IupDrawEnd(ih);
  return IUP_DEFAULT;
}

int val_valuechanged_cb(Ihandle *self)
{
  Ihandle* dial = IupGetDialog(self);
  Ihandle* canvas = (Ihandle *)IupGetAttribute(self, "CANVAS");
  Ihandle *undoButton = (Ihandle *)IupGetDialogChild(self, "UNDO");
  Ihandle *redoButton = (Ihandle *)IupGetDialogChild(self, "REDO");
  int circleId = IupGetInt(dial, "CIRCLEID");
  double value = IupGetDouble(self, "VALUE");
  Circle circle = get(circleId);
  int rOriginal = circle.rOriginal;

  update(circleId, rOriginal*value);

  addOperation(CHANGE_RADIUS, get(circleId));

  if (undoLeft())
    IupSetAttribute(undoButton, "ACTIVE", "YES");
  else
    IupSetAttribute(undoButton, "ACTIVE", "NO");

  if (redoLeft)
    IupSetAttribute(redoButton, "ACTIVE", "YES");
  else
    IupSetAttribute(redoButton, "ACTIVE", "NO");

  IupUpdate(canvas);

  return IUP_DEFAULT;
}

Ihandle *createCircleConfigDial()
{
  Ihandle *dial, *val, *vbox, *box;

  val = IupVal("HORIZONTAL");
  IupSetAttribute(val, "EXPAND", "HORIZONTAL");
  IupSetAttribute(val, "MIN", "0.5");
  IupSetAttribute(val, "MAX", "1.5");
  IupSetAttribute(val, "VALUE", "1");

  box = IupVbox(IupFill(), val, IupFill(), NULL);
  IupSetAttribute(box, "ALIGMENT", "ACENTER");

  IupSetAttribute(box, "NMARGIN", "10x10");
  dial = IupDialog(box);
  IupSetAttribute(dial, "SIZE", "100x60");

  IupSetCallback(val, "VALUECHANGED_CB", (Icallback)val_valuechanged_cb);

  return dial;
}

int main(int argc, char **argv)
{
  Ihandle *dlg, *hbox, *vbox, *configDial;
  Ihandle *undoButton, *redoButton, *canvas;

  IupOpen(&argc, &argv);

  undoButton = IupButton("Undo", NULL);
  IupSetAttribute(undoButton, "NAME", "UNDO");
  IupSetAttribute(undoButton, "SIZE", "60");
  redoButton = IupButton("Redo", NULL);
  IupSetAttribute(redoButton, "NAME", "REDO");
  IupSetAttribute(redoButton, "SIZE", "60");

  IupSetAttribute(redoButton, "ACTIVE", "NO");
  IupSetAttribute(undoButton, "ACTIVE", "NO");

  canvas = IupCanvas(NULL);
  IupSetAttribute(canvas, "NAME", "CANVAS");
  IupSetAttribute(canvas, "EXPAND", "YES");

  hbox = IupHbox(IupFill(), undoButton, redoButton, IupFill(), NULL);

  vbox = IupVbox(hbox, canvas, NULL);

  IupSetAttribute(vbox, "NMARGIN", "10x10");
  IupSetAttribute(vbox, "GAP", "10");
  dlg = IupDialog(vbox);
  IupSetAttribute(dlg, "TITLE", "Circle Drawer");
  IupSetAttribute(dlg, "SIZE", "300x150");

  /* Registers callbacks */
  IupSetCallback(undoButton, "ACTION", (Icallback)bt_undo_cb);
  IupSetCallback(redoButton, "ACTION", (Icallback)bt_redo_cb);
  IupSetCallback(canvas, "ACTION", (Icallback)action);
  IupSetCallback(canvas, "BUTTON_CB", (Icallback)button_cb);
  IupSetCallback(canvas, "MOTION_CB", (Icallback)motion_cb);

  configDial = createCircleConfigDial();

  IupSetAttribute(configDial, "CANVAS", (char *)canvas);
  IupSetAttribute(configDial, "UNDO", (char *)undo);
  IupSetAttribute(configDial, "REDO", (char *)redo);

  IupSetAttribute(dlg, "CONFIGDIAL", (char *)configDial);

  IupShowXY(dlg, IUP_CENTER, IUP_CENTER);

  IupMainLoop();

  IupClose();
  return EXIT_SUCCESS;
}

File: iup/html/en/7gui/circle_drawer.lua
require("iuplua")

opType = { }
opType.create = 1
opType.remove = 2
opType.update = 3

operations = {}
currentOperationId = 0
operationsCount = 0

function op_add(operation, id, x, y, radius)
	local opr= {}
	if currentOperationId < operationsCount then
		operationsCount = currentOperationId
	end
	opr.operation = operation
	opr.id = id
	opr.x = x
	opr.y = y
	opr.radius = radius
	table.insert(operations, currentOperationId+1, opr)
	operationsCount = operationsCount + 1
	currentOperationId = operationsCount
end

function op_has_undo()
	if currentOperationId < 1 then
		return nil
	end
	return 1
end

function op_has_redo()
	if currentOperationId == operationsCount then
		return nil
	end
	return 1
end

function op_undo()
	if currentOperationId < 1 then
		return
	end
	local operation = operations[currentOperationId]
	if  operation.operation == opType.create then
		circles_remove(operation.id)
	elseif operation.operation == opType.remove then
		circles_update(operation.id, operation.x, operation.y, operation.radius)
	else
		circles_update(operation.id, operation.x, operation.y, operation.radius)
	end
	currentOperationId = currentOperationId - 1
end

function op_redo()
	if currentOperationId == operationsCount then
		return
	end
	currentOperationId = currentOperationId + 1
	local operation = operations[currentOperationId]
	if  operation.operation == opType.create then
		circles_add(operation.x, operation.y, operation.id)
	elseif operation.operation == opType.remove then
		circles_update(operation.id, operation.x, operation.y, operation.radius)
	else
		circles_update(operation.id, operation.x, operation.y, operation.radius)
	end
end

DEF_RADIUS = 30

circles = {}
lastId = 0

function circles_add(x, y, id)
	local circle = {}
	circle.x = x
	circle.y = y
	circle.radius = DEF_RADIUS
	local validId
	if not id then
		lastId = lastId + 1
		validId = lastId
	else
		validId = id
	end
	circle.id = validId
	circles[validId] = circle
	return lastId
end

function circles_find(id)
	return circles[id].x, circles[id].y, circles[id].radius
end

function circles_remove(id)
	circles[id] = nil
end

function circles_update(id, x, y, radius)
	circles[id].x = x
	circles[id].y = y
	circles[id].radius = radius
end

function circles_pick(x, y)
	local i, dist, dx, dy, k, v
	local selectedCircle = -1
	for k, v in pairs(circles) do
		local dx = v.x - x
		local dy = v.y - y
		local d = ((dx*dx) + (dy*dy))^.5
		if  d < v.radius then
			if  selectedCircle < 0 then
				selectedCircle = v.id
				dist = d
			elseif d < dist then
				selectedCircle = v.id
			end
		end
	end
	return selectedCircle
end


--********************************** Main *****************************************

undoButton = iup.button{title = "Undo", size = "60", active = "NO"}
redoButton = iup.button{title = "Redo", size = "60", active = "NO"}

canvas = iup.canvas{expand = "YES", highlightedCircle = -1}

hbox = iup.hbox{iup.fill{}, undoButton, redoButton, iup.fill{}}

vbox = iup.vbox{hbox, canvas;  nmargin = "10x10", gap = "10"}

dlg = iup.dialog{vbox, title = "Circle Drawer", size = "300x150"}

val = iup.val{"HORIZONTAL"; expand = "HORIZONTAL", min = "0.5", max = "1.5", value = "1"}

config_box = iup.vbox{iup.fill{}, val, iup.fill{}, alignment = "ACENTER", nmargin = "10x10"}

config_dial = iup.dialog{config_box; size = "100x60"}

function undoButton:action()
	op_undo()
	if op_has_undo() then
		undoButton.active = "YES"
	else
		undoButton.active = "NO"
	end
	if op_has_redo() then
		redoButton.active = "YES"
	else
		redoButton.active = "NO"
	end
	iup.Update(canvas);
	return iup.default
end

function redoButton:action()
	op_redo()
	if op_has_undo() then
		undoButton.active = "YES"
	else
		undoButton.active = "NO"
	end
	if op_has_redo() then
		redoButton.active = "YES"
	else
		redoButton.active = "NO"
	end
	iup.Update(canvas);
	return iup.default
end

function drawCircles(canvas)
	local i, w, h

	iup.DrawBegin(canvas)

	w, h = iup.DrawGetSize(canvas)

	iup.SetAttribute(canvas, "DRAWCOLOR", "255 255 255")
	iup.SetAttribute(canvas, "DRAWSTYLE", "FILL")
	iup.DrawRectangle(canvas, 0, 0, w - 1, h - 1)

	iup.SetAttribute(canvas, "DRAWCOLOR", "0 0 0")
	iup.SetAttribute(canvas, "DRAWSTYLE", "STROKE")
	canvas["DRAWCOLOR"] = "0 0 0"
	canvas["DRAWSTYLE"] = "STROKE"

	for k, v in pairs(circles) do
		local circleId = v.id
		if  tonumber(canvas.highlightedCircle) == circleId then
			iup.SetAttribute(canvas, "DRAWCOLOR", "128 128 128")
			iup.SetAttribute(canvas, "DRAWSTYLE", "FILL")
		else
			iup.SetAttribute(canvas, "DRAWCOLOR", "0 0 0")
			iup.SetAttribute(canvas, "DRAWSTYLE", "STROKE")
		end

		iup.DrawArc(canvas, v.x - v.radius, v.y - v.radius,
					v.x + v.radius, v.y + v.radius, 0., 360.)
	end

	iup.DrawEnd(canvas)
end

function canvas:action(posx, posy)
	drawCircles(self)
	return iup.default
end

function canvas:button_cb(button, pressed, x, y, status)
	local id
	if  button == iup.BUTTON1 and pressed == 1 then
		id = circles_add(x, y)
		op_add(opType.create, id, x, y, DEF_RADIUS)

		if op_has_undo() then
			undoButton.active = "YES"
		else
			undoButton.active = "NO"
		end

		if op_has_redo() then
			redoButton.active = "YES"
		else
			redoButton.active = "NO"
		end
	
	elseif  button == iup.BUTTON3 and pressed == 1 then
		id = circles_pick(x, y)
		if id >= 0 then
			config_dial.circleId = id
			iup.Popup(config_dial, iup.CENTER, iup.CENTER)
			self.highlightedCircle = -1
		end
	end
	iup.Update(self)
	return iup.default
end

function canvas:motion_cb(x, y, status)
	self.highlightedCircle = circles_pick(x, y)
	iup.Update(self)
	return iup.default
end

function val:valuechanged_cb()
	local dial = iup.GetDialog(self)
	local value = self.value
	local circleId = tonumber(dial.circleId)
	local x, y, radius = circles_find(circleId)
	circles_update(circleId, x, y, DEF_RADIUS*value)
	op_add(opType.update, circleId, x, y, radius)

	if op_has_undo() then
		undoButton.active = "YES"
	else
		undoButton.active = "NO"
	end

	if op_has_redo() then
		redoButton.active = "YES"
	else
		redoButton.active = "NO"
	end

	iup.Update(canvas)
	
	return iup.default
end

dlg:showxy( iup.CENTER, iup.CENTER )

if (iup.MainLoopLevel()==0) then
	iup.MainLoop()
end
